src/Components/Components/src/Reflection/ComponentProperties.cs

```diff
@@ -56,13 +56,22 @@ public static void SetProperties(in ParameterView parameters, object target)
                     ThrowForSettingCascadingParameterWithNonCascadingValue(targetType, parameterName);
                     throw null; // Unreachable
                 }
-                else if (!writer.Cascading && parameter.Cascading)
+                else if (!writer.AcceptsCascadingParameters && parameter.Cascading)
                 {
                     // We're giving a more specific error here because trying to set a non-cascading parameter
                     // with a cascading value is likely deliberate (but not supported), or is a bug in our code.
                     ThrowForSettingParameterWithCascadingValue(targetType, parameterName);
                     throw null; // Unreachable
                 }
+                else if (parameter.Cascading && writer.AcceptsDirectParameters && writer.AcceptsCascadingParameters)
+                {
+                    // It's possible that the writer accepts both cascading and direct parameters. In that case,
+                    // we want to prefer the direct parameter, if it exists.
+                    if (parameters.HasDirectParameter(parameterName))
+                    {
+                        continue;
+                    }
```

Comment:
This is way more general behavior than what I expected.

In my view, we should just see where we are throwing the exception and special case it for supply from query parameters. We are doing this for back compat, we are not trying to create a new language/idiom.

Reply:
I agree. If we are able to limit this to just backcompat, I think we'll end up in a better place where we don't have the nonobviousness of needing priority rules to pick between cascading and direct parameters. I could be missing something, but perhaps the only change we need is:

 * When a direct parameter arrives, whether or not it matches a `[Parameter]` property, we continue to check if there's also `[SomeCascadingParameter]` on the target property. If there is:
   * **New behavior:** If it's `[SupplyParameterFromQuery]` and there is also `[Parameter]` on that property, we allow it and prefer the direct value for back-compat
   * Otherwise we throw like we did before

That might be the only case we have to deal with. It would mean that:

  * `[Parameter] [SupplyParameterFromQuery]` would accept both direct and cascaded values, preferring direct
  * `[Parameter] [SomeOtherCascadedParameter]` would never be useful, because it would throw if a direct parameter was received
  * `[AnyCascadedParameter]` on its own, including `[SupplyParameterFromQuery]`, could only receive cascaded values for the same reason (if a direct value was received at runtime, it would throw)

Does that sound right?

Reply:
That analysis seems right to me. My initial reasoning for not limiting the fix to just `[SupplyParameterFromQuery]` was that it felt weird to have something so special-cased in the core parameter setting logic. But I do acknowledge that it's a safer to not introduce fundamental new behavior in that area.

> When a direct parameter arrives and matches a `[Parameter]` property, we continue to check if there's also `[SomeCascadingParameter]` on the target property. If there is...

Yep, this logic seems right to me and is equivalent to what I've implemented (AFAICT), minus the part about checking the existence of `[SupplyParameterFromQuery]` specifically. In my commit addressing this feedback, the core parameter setting code (`ComponentProperites.SetProperties()`) is still attribute-agnostic, but I made a one-line change to `WritersForType` so that properties can only receive direct parameters if they don't have a `CascadingParameterAttribute`, unless that attribute is a `SupplyParameterFromQueryAttribute`.